### ISSUE TITLE
Memory problem when reloading the model in parallel inference

### ISSUE THREAD

**Issue by icechen** [deeplearning4j/deeplearning4j#4328]
### Issue Description
In our application the ComputationGraph model is re-trained and reloaded regularly. We found reloading the models (which we run using ParallelInference) repeatedly would exhaust GPU RAM before the program stops responding.

The example code to reproduce the problem:

```Java
String modelFileName = "MODEL_FILE_NAME";
int cnt=0;
while (true) {
    System.out.println(String.format("Model reloaded for %d times", ++cnt));
    ComputationGraph net = ModelSerializer.restoreComputationGraph(modelFileName, true);
    ParallelInference parallelNet = new ParallelInference
            .Builder(net)
            .inferenceMode(InferenceMode.SEQUENTIAL)
            .workers(4)
            .build();
    Thread.sleep(3000);
}
```

### Version Information
Server environment:
* OS: ubuntu 14.04
* CPU: i7-6850K 3.60GHz
* RAM: 64G
* Graphic card: NVIDIA TITAN X * 4
* Java: 1.8.0_144
* Maven: 3.5.0
* Deeplearning4j version: 0.9.1
* CUDA version: nd4j-cuda-8.0-platform
* NVIDIA driver version: 375.66

### Contributing
The cause seems to be un-released InferenceWorker threads, which keep waiting for the inputs in the while loop:

In ParallelInference.java:
```Java
while (shouldWork.get()) {
                    InferenceObservable request = inputQueue.take();
```

This prevents the thread from exiting the loop if there is no input. So every time we new a ParallelInference object, we make 3 more replicated model in the above example code and accumulate the GPU RAM consumption.

Our temporary solution:

1. Set timeout when taking inputs from BlockingQueue:
    ```Java
    InferenceObservable request = inputQueue.poll(10, TimeUnit.MILLISECONDS);
    if (request != null) {
    	//do prediction
    }
    ```
2. Write a shutdown function in ParallelInference (by extending ParallelInference class)
    ```Java
    public void shutdown() {
        if (zoo != null) {
            for (InferenceWorker worker: zoo) {
                worker.shutdown();
            }
        }
    }
    ```
3. Call the shutdown function every time before reloading the ComputationGraph model

**Comment by agibsonccc** [comment]
Closing — this issue is over 7 years old with no recent activity. The codebase has undergone a major rewrite. If this is still relevant, please open a new issue with current reproduction steps.
